kanidmd/src/lib/idm/authsession.rs

```diff
@@ -301,6 +309,23 @@ impl CredHandler {
                             CredState::Denied(BAD_TOTP_MSG)
                         }
                     }
+                    (AuthCredential::BackupCode(code_chal), _, _, Some(backup_code)) => {
```

Comment:
You need to extend and test this state machine flow in the tests at the bottom of this file. This is a security critical code path, so requires *extensive* testing. However, the logic here appears to be correct IMO.